### PR TITLE
[Feat] implement calculator with some unit testing

### PR DESCRIPTION
**Pull request by david-wagih**

### CODE REVIEW

File: test_calculator.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import unittest
+from calculator import Calculator
+
+class TestCalculator(unittest.TestCase):
+    def setUp(self):
+        self.calc = Calculator()
```
Comment (david-wagih) [inline test_calculator.py:6]: [INFO] The setUp method should be named setup for consistency with Python naming conventions.

File: test_calculator.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+    def test_calculate_square(self):
+        self.assertEqual(self.calc.calculate_square(4), 16)
```
Comment (david-wagih) [inline test_calculator.py:9]: [WARNING] Consider adding more tests to cover different edge cases and scenarios. This will help ensure the robustness of the code.

File: test_calculator.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        with self.assertRaises(ValueError):
+            self.calc.calculate_square_root(-1)
+
+    def test_history_management(self):
+        # Test adding to history
+        self.calc.add_to_history(5)
+        self.assertEqual(len(self.calc.calculation_history), 1)
+        
+        # Test history size limit
+        for i in range(15):
```
Comment (david-wagih) [inline test_calculator.py:27]: [ERROR] Ensure that all exceptions are handled properly in your unit tests. Currently, some expected errors are not being tested.

File: main.py [hunk 4]
```diff
@@ -1,6 +1,30 @@
+from calculator import Calculator
+
+
 def main():
```
Comment (david-wagih) [inline main.py:4]: [INFO] Imports should be at the top of the file.

File: main.py [hunk 5]
```diff
@@ -1,6 +1,30 @@
...
-    print("Hello, World!")
+    calculator = Calculator()
+    
+    # Get user input and perform calculations
```
Comment (david-wagih) [inline main.py:7]: [WARNING] The function 'main' is too short and does not warrant its own separate function. Consider inline code if it simplifies the main body significantly.

File: main.py [hunk 6]
```diff
@@ -1,6 +1,30 @@
...
+    number = float(get_user_input())
+    
+    # Perform various calculations
```
Comment (david-wagih) [inline main.py:10]: [INFO] Use a more meaningful variable name than 'number'. Naming conventions should be followed for better readability and maintainability.

File: main.py [hunk 7]
```diff
@@ -1,6 +1,30 @@
...
+    print(f"Square: {calculator.calculate_square(number)}")
+    print(f"Cube: {calculator.calculate_cube(number)}")
+    print(f"Square root: {calculator.calculate_square_root(number)}")
+    
+    # Store calculation history
```
Comment (david-wagih) [inline main.py:15]: [WARNING] Consider breaking down the calculation methods into their respective functions or classes to adhere to the Single Responsibility Principle. Each method should have a single responsibility.

File: main.py [hunk 8]
```diff
@@ -1,6 +1,30 @@
...
+    calculator.add_to_history(number)
+    print("\nCalculation History:")
+    calculator.display_history()
+
+def get_user_input():
+    while True:
+        try:
+            return float(input("Enter a positive number: "))
```
Comment (david-wagih) [inline main.py:23]: [INFO] The 'get_user_input' function can be simplified and inlined if not reused elsewhere.

File: main.py (hunk 6 above)
Comment (david-wagih) [inline main.py:10]: [WARNING] User input should be validated to ensure it is a positive number before performing calculations. Consider using a while loop with an appropriate condition to enforce this.

File: calculator.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
+import math
+
+class Calculator:
+    def __init__(self):
+        self.calculation_history = []
+        self.max_history_size = 10
```
Comment (david-wagih) [inline calculator.py:6]: [WARNING] The `max_history_size` attribute is hardcoded and should be made configurable or documented as a constant.

File: calculator.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+    def calculate_square(self, number):
+        """Calculate the square of a number."""
+        result = number * number
+        return result
+
+    def calculate_cube(self, number):
```
Comment (david-wagih) [inline calculator.py:13]: [INFO] The function name 'calculate_square' does not clearly describe its action. Consider renaming it to 'square' for better clarity.

File: calculator.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+        """Calculate the cube of a number."""
+        return number ** 3
+
+    def calculate_square_root(self, number):
+        """Calculate the square root of a number."""
```
Comment (david-wagih) [inline calculator.py:18]: [INFO] The function name 'calculate_cube' is clear but consider adding a docstring to explain the method further.

File: calculator.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        if number < 0:
+            raise ValueError("Cannot calculate square root of negative number")
+        return math.sqrt(number)
+
+    def add_to_history(self, number):
```
Comment (david-wagih) [inline calculator.py:23]: [WARNING] The function 'calculate_square_root' should include proper error handling for negative numbers and possibly other edge cases. Consider raising an exception if the number is invalid.

File: calculator.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+        """Add a number to calculation history."""
+        self.calculation_history.append(number)
+        if len(self.calculation_history) > self.max_history_size:
+            self.calculation_history.pop(0)
```
Comment (david-wagih) [inline calculator.py:27]: [ERROR] A docstring is missing for the '__init__' method which should explain the purpose of initializing calculation history and max_history_size.

File: calculator.py [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+    def display_history(self):
+        """Display the calculation history."""
+        for i, num in enumerate(self.calculation_history, 1):
+            print(f"Entry {i}: {num}")
+
+    def clear_history(self):
+        """Clear the calculation history."""
+        self.calculation_history = []
```
Comment (david-wagih) [inline calculator.py:36]: [ERROR] The function 'clear_history' should be documented. It is currently undocumented which could lead to confusion about its purpose and usage.

**Review by david-wagih — COMMENTED** [review]
Code Review Feedback

Automated review by MF Code Reviewer

File: calculator.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
+import math
+
+class Calculator:
+    def __init__(self):
+        self.calculation_history = []
+        self.max_history_size = 10
+
```
Comment (david-wagih) [inline calculator.py:7]: [WARNING] Consider using a more descriptive method name instead of `calculate_square` for methods that perform operations.

File: calculator.py [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+    def calculate_square(self, number):
+        """Calculate the square of a number."""
+        result = number * number
+        return result
+
```
Comment (david-wagih) [inline calculator.py:12]: [WARNING] The current implementation of the square root method does not handle complex numbers. It would be safer to use the `cmath` module which handles complex numbers.

File: calculator.py [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+    def calculate_cube(self, number):
+        """Calculate the cube of a number."""
+        return number ** 3
```
Comment (david-wagih) [inline calculator.py:15]: [WARNING] The method for adding to history should be more robust against non-numeric values.

File: calculator.py [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+
+    def calculate_square_root(self, number):
+        """Calculate the square root of a number."""
+        if number < 0:
```
Comment (david-wagih) [inline calculator.py:19]: [WARNING] The history display method lacks error handling for when there's nothing in the history.

File: calculator.py [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+
+    def add_to_history(self, number):
+        """Add a number to calculation history."""
+        self.calculation_history.append(number)
+        if len(self.calculation_history) > self.max_history_size:
+            self.calculation_history.pop(0)
+
+    def display_history(self):
+        """Display the calculation history."""
+        for i, num in enumerate(self.calculation_history, 1):
```
Comment (david-wagih) [inline calculator.py:31]: [INFO] The method to clear history does not return anything and lacks comments on its behavior.

File: calculator.py [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+            print(f"Entry {i}: {num}")
+
+    def clear_history(self):
+        """Clear the calculation history."""
```
Comment (david-wagih) [inline calculator.py:35]: [INFO] There's no need for a newline character at the end of the file as Python automatically ignores them.

File: test_calculator.py [hunk 21]
```diff
@@ -0,0 +1,32 @@
+import unittest
+from calculator import Calculator
+
+class TestCalculator(unittest.TestCase):
+    def setUp(self):
+        self.calc = Calculator()
+
```
Comment (david-wagih) [inline test_calculator.py:7]: [INFO] Consider adding a docstring to the TestCalculator class and its methods for better documentation.

File: test_calculator.py [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+    def test_calculate_square(self):
```
Comment (david-wagih) [inline test_calculator.py:8]: [INFO] Ensure that setUp is necessary. If you don't use it in other methods, consider removing it to keep the code clean.

File: test_calculator.py [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+        self.assertEqual(self.calc.calculate_square(4), 16)
+        self.assertEqual(self.calc.calculate_square(-2), 4)
+
+    def test_calculate_cube(self):
+        self.assertEqual(self.calc.calculate_cube(3), 27)
+        self.assertEqual(self.calc.calculate_cube(-2), -8)
+
+    def test_calculate_square_root(self):
```
Comment (david-wagih) [inline test_calculator.py:16]: [WARNING] The test_calculate_square method is missing a case where the input is zero. It should raise an error according to the function's logic.

File: test_calculator.py [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+        self.assertEqual(self.calc.calculate_square_root(16), 4)
+        with self.assertRaises(ValueError):
+            self.calc.calculate_square_root(-1)
+
+    def test_history_management(self):
+        # Test adding to history
+        self.calc.add_to_history(5)
+        self.assertEqual(len(self.calc.calculation_history), 1)
+
```
Comment (david-wagih) [inline test_calculator.py:25]: [WARNING] The test_history_management method lacks a clear assertion to verify the history content after adding multiple entries.

File: test_calculator.py [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+    def test_history_management(self):
+        # Test adding to history
+        self.calc.add_to_history(5)
+        self.assertEqual(len(self.calc.calculation_history), 1)
+        
+        # Test history size limit
+        for i in range(15):
+            self.calc.add_to_history(i)
+        self.assertEqual(len(self.calc.calculation_history), 10)
+
```
Comment (david-wagih) [inline test_calculator.py:30]: [WARNING] The line `self.assertEqual(len(self.calc.calculation_history), 10)` might not be appropriate if the history management logic changes.

File: test_calculator.py [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+if __name__ == '__main__':
+    unittest.main()
```
Comment (david-wagih) [inline test_calculator.py:32]: [WARNING] Missing a newline at the end of the file. Python requires this for some contexts.

File: main.py [hunk 27]
```diff
@@ -1,6 +1,30 @@
+from calculator import Calculator
+
+
 def main():
-    print("Hello, World!")
+    calculator = Calculator()
```
Comment (david-wagih) [inline main.py:5]: [INFO] Consider adding a docstring to the `main` function to explain its purpose and provide context for future maintainers.

File: main.py [hunk 28]
```diff
@@ -1,6 +1,30 @@
...
+
```
Comment (david-wagih) [inline main.py:6]: [INFO] The import statement should be placed after any module-level constants or functions to follow Python's style guide.

File: main.py (hunk 6 above)
Comment (david-wagih) [inline main.py:10]: [INFO] Using a float for this kind of input is unconventional. Typically, you would accept an integer or use a string to avoid potential floating-point errors.

File: main.py [hunk 29]
```diff
@@ -1,6 +1,30 @@
...
+    calculator = Calculator()
+    
+    # Get user input and perform calculations
+    number = float(get_user_input())
+    
+    # Perform various calculations
+    print(f"Square: {calculator.calculate_square(number)}")
+    print(f"Cube: {calculator.calculate_cube(number)}")
+    print(f"Square root: {calculator.calculate_square_root(number)}")
+
```
Comment (david-wagih) [inline main.py:14]: [WARNING] The current implementation of `get_user_input` does not handle exceptional cases effectively. It should have more robust error handling.

File: main.py [hunk 30]
```diff
@@ -1,6 +1,30 @@
...
+    
+    # Perform various calculations
+    print(f"Square: {calculator.calculate_square(number)}")
+    print(f"Cube: {calculator.calculate_cube(number)}")
+    print(f"Square root: {calculator.calculate_square_root(number)}")
+    
+    # Store calculation history
+    calculator.add_to_history(number)
+    print("\nCalculation History:")
+    calculator.display_history()
```
Comment (david-wagih) [inline main.py:18]: [WARNING] The `calculator` object should be instantiated in the main function and passed around as needed. This will make it easier to test and avoid potential issues with state.

File: main.py [hunk 31]
```diff
@@ -1,6 +1,30 @@
...
+
+def get_user_input():
+    while True:
+        try:
+            return float(input("Enter a positive number: "))
+        except ValueError:
```
Comment (david-wagih) [inline main.py:24]: [ERROR] The method `add_to_history` and `display_history` are not defined in the provided `calculator.py`. Ensure these methods exist or consider removing them from the code.

File: main.py [hunk 32]
```diff
@@ -1,6 +1,30 @@
...
+            print("Please enter a valid number.")
 
 if __name__ == "__main__":
     main()
+
```
Comment (david-wagih) [inline main.py:29]: [INFO] The condition `if __name__ == "__main__":` is unnecessary as it only contains function calls. Remove this conditional to simplify the script.

**Review by david-wagih — COMMENTED** [review]
Code Review Feedback

Automated review by MF Code Reviewer
